### PR TITLE
Features/methods

### PR DESCRIPTION
**Pull request by jacobo12montes**
we re-implemented ruby enumerable methods. Based on the documentation of every method we mimicked the behavior of that is described in the documentation without reusing the original methods. as part of the behavior, we defined our methods inside the Enumerable module to be able to call it on arrays in the same way as the original ones.

### CODE REVIEW

File: methods.rb [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  end
+end
+
+def multiply_els(array)
+  array.my_inject(:*)
+end
+# proc_map = Proc.new { |i| i * i }
+# p [1,2,3].my_map(proc_map) {|i| i + i}
+# p multiply_els([2,4,5])
+# p [1,2,3].my_inject
```
Comment (adaorachi) [inline methods.rb]: - [ ] Kindly remove the commented code. If you need to review a previous version of your code, you can use git history.

**Comment by adaorachi** [comment]
### Hi @jacobo12montes and @mohamedSabry0 :wave: 

Well done on the completion of the first draft of this project. However, your enumerable methods are not passing a handful of test cases.
Kindly read through my comments below and see the changes to implement. 
**There are a lot of test cases your enumerable methods are not passing, so you have to patiently go through the failed test cases below and implement them in your methods.**

--- 

### CHANGES REQUIRED :recycle: 

---

### After implementing the changes, you can request for another review by submitting a code review form.

Should you have any questions, kindly reach out to me on Slack with the username - **@adaorachi**

Happy Coding :relaxed: 

**MaryAnn Chukwuka**

**Review by adaorachi — CHANGES_REQUESTED** [review]

**Review by adaorachi — CHANGES_REQUESTED** [review]
- [x] Code not related to the project should not be committed but added to .gitignore, e.g - my_inject.txt.

- [ ] **- The following methods did not output the same result when compared with their equivalent Ruby methods. Kindly ensure you fix them.**

-  **#my_each method** 
   - [x]  when a block is given and when self is a range. It should return the range itself after calling the given block once for each element in self. 
`expect Range.new(1, 5).my_each(&block) to equal Range.new(1, 5).each(&block)`.
   - [x] when a block is given and when self is a hash, it should behave like the Ruby #each and calls the given block once for each element in self. 
`expect { one: "1", two: "2" }.my_each(&proc { |w| puts w }) to equal { one: "1", two: "2" }.each(&proc { |w| puts w })`

<br />

- **#my_each_with_index method**
  - [x] when a block is given and when self is a range, it should return the range itself after calling the given block once for each element in self. 
` expect Range.new(1, 5).my_each_with_index(&block) to equal Range.new(1, 5).each_with_index(&block)`.
  - [x] when a block is given and when self is a hash, it should behave like Ruby's #each_with_index method and calls the given block once for each element in self.
`expects { one: "1", two: "2" }.my_each_with_index(&proc { |x,y| puts x, y }) to equal { one: "1", two: "2" }.each_with_index(&proc { |x,y| puts x, y })`

<br />

- **#my_select? method**
  - [x] returns an array containing all elements of enum for which the given block returns a true value
    `expect Range.new(1, 5).my_select(&block) to equal (Range.new(1, 5).select(&block)`
  - [x] returns false if the block returns false or nil::range
    `expect Range.new(1, 5).my_all?(&false_block)) to equal range.all?(&false_block)`

<br />

- **#my_all?**
  - [x]  when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
       ` expect [1,2,3].all?(/d/).my_all?(Integer)) to equal [1,2,3].all?(Integer)`
  - [x] when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
      ` expect ['dog', 1, 'rod'].all?(/d/).my_all?(Numeric)) to equal ['dog', 1, 'rod'].all?(Numeric)`
  - [x] when a class is passed as an argument returns false if any of the collection is not a member of such class
     ` expect ['dog', 1, 'rod'].all?(/d/).my_all?(Integer)) to equal ['dog', 1, 'rod'].all?(Integer)`
  - [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
     `expect %w[dog door rod].my_all?(/d/)) to equal %w[dog door rod].all?(/d/)`
  - [x] when a Regex is passed as an argument returns false if any of the collection does not match the Regex
     `expect %w[dog door rod].my_all?(/o/) should equal %w[dog door rod].all?(/o/)`
  - [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
     `expect [3,3,3].my_all?(3) to equal [3,3,3].all?(3)`
  - [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
      `expect %w[dog door rod].my_all?('rod') to equal %w[dog door rod].all?('rod')`

<br />

- **my_any? method**
  - [x]  returns false if the block ever returns a value false or nil::range
     `expect range.my_any?(&false_block) to equal range.any?(&false_block)`
  - [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
      `expect ['dog', 1, 'door'].my_any?(Numeric) to equal ['dog', 1, 'door'].any?(Numeric)`
  - [x]  when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
     `expect ['dog', 1, 'door'].my_any?(Integer) to equal ['dog', 1, 'door'].any?(Integer)`
  - [x] when a Regex is passed as an argument returns true if any of the collection matches the Regex
      `expect %w[dog door rod].my_any?(/d/) to equal %w[dog door rod].any?(/d/)`
  - [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
      `expect %w[dog door rod].my_any?(/z/) to equal %w[dog door rod].any?(/z/)`
   - [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
     `expect %w[dog door rod].my_any?('cat') to equal  %w[dog door rod].any?('cat')`
  - [x] when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
    `expect %w[dog door rod].my_any?('rod') to equal  %w[dog door rod].any?('rod')`

<br />

- **#my_none? method** 
  - [x] returns false if the block ever returns true for all elements - **for range**
     ` expect range.my_none?(&false_block) to equal range.none?(&false_block)`
  - [x] when a class is passed as an argument returns true if none of the collection is a member of such class
      `expect [1,2,3].my_none?(String)) to equal [1,2,3].my_none?(String)`
  - [x] when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
      expect %w[dog door rod].my_none?(Numeric)) to equal %w[dog door rod].my_none?(Numeric)
  - [x] when a class is passed as an argument returns false if any of the collection is a member of such class
    ` expect %w[dog door rod].my_none?(String)) to equal %w[dog door rod].my_none?(String)`
  - [x]  when a Regex is passed as an argument returns true if none of the collection matches the Regex
     `expect %w[dog door rod].my_none?(/z/)) to equal %w[dog door rod].none?(/z/)`
  - [x] when a Regex is passed as an argument returns false if any of the collection matches the Regex
     `expect %w[dog door rod].my_none?(/d/) to equal %w[dog door rod].none?(/d/)`
  - [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
      `expect  %w[dog door rod].none?('rod') to equal  %w[dog door rod].my_none?('rod')`
  - [x] when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
     ` expect %w[dog door rod].my_none?('cat') to equal  %w[dog door rod].none?('cat')`

<br />

- **#my_count method** 
  - [x] returns the number of items in enum through enumeration - **for range**
      `expect Range.new(1, 5).my_count to equal Range.new(1, 5).count`

<br />

 - **#my_map method** 
   - [ ] returns a new array with the results of running a given block once for every element in enum **- for range**
     `expect [1,2,3].my_map(&block) to equal [1,2,3].map(&block)`

<br />

- **#my_inject method** 
  - [x] raises a "LocalJumpError" when no block or argument is given
    `expect [1,2,3].my_inject to raise_error - (LocalJumpError)`
  - [x] when a block is given without an initial value, it combines all elements of the enum by applying a binary operation, and specified by a block- **for range**
  `expect Range.new(1, 5).inject { |prod, n| prod * n } should equal Range.new(1, 5).my_inject { |prod, n| prod * n }`
  - [x] when a block is given with an initail value, it combines all elements of the enum by applying a binary operation and specified by a block
     `expect Range.new(1, 5).inject(4) { |prod, n| prod * n } should equal Range.new(1, 5).my_inject(4) { |prod, n| prod * n }`
  - [x] when a symbol is specified without an initail value combines each element of the collection by applying the symbol as a named method - **for range**
    ` expect Range.new(1, 5).my_inject(:*) to equal Range.new(1, 5).inject(:*)`
  - [x]  when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method - **for range**
     `expect Range.new(1, 5).my_inject(2, :*) to equal Range.new(1, 5).inject(2, :*)`

**Review by sumancrest0001 — APPROVED** [review]
**Status: Approved :100: :1st_place_medal: 
Congratulations team :clap:** 
You have done an exceptional job to make all the enumerable methods similar to the default methods in functionality. Now you can merge the branch and submit the project completion form. Wish you all the best with the next project.
**[NOTE]: We should write commit messages in the present tense and remember while working on future projects.**
Happy coding !!!
mission

**Comment by mohamedSabry0** [comment]
I'm submitting another code review request as my dashboard's version is different from my partner's. And as a result, my project status didn't change yet. I tried to contact support but no answer yet.

**Pull request merged**